packages/perseus-core/src/utils/util.ts

```diff
@@ -21,6 +21,54 @@ const seededRNG: (seed: number) => RNG = function (seed: number): RNG {
     };
 };
 
+// Note: Previously called cyrb128
+function getHashes(str: string): [number, number, number, number] {
+    let h1 = 1779033703;
+    let h2 = 3144134277;
+    let h3 = 1013904242;
+    let h4 = 2773480762;
+    let k = 0;
+    for (let i = 0; i < str.length; i++) {
+        k = str.charCodeAt(i);
+        h1 = h2 ^ Math.imul(h1 ^ k, 597399067);
+        h2 = h3 ^ Math.imul(h2 ^ k, 2869860233);
+        h3 = h4 ^ Math.imul(h3 ^ k, 951274213);
+        h4 = h1 ^ Math.imul(h4 ^ k, 2716044179);
+    }
+    h1 = Math.imul(h3 ^ (h1 >>> 18), 597399067);
+    h2 = Math.imul(h4 ^ (h2 >>> 22), 2869860233);
+    h3 = Math.imul(h1 ^ (h3 >>> 17), 951274213);
+    h4 = Math.imul(h2 ^ (h4 >>> 19), 2716044179);
+    return [
+        (h1 ^ h2 ^ h3 ^ h4) >>> 0,
+        (h2 ^ h1) >>> 0,
+        (h3 ^ h1) >>> 0,
+        (h4 ^ h1) >>> 0,
+    ];
+}
+
+// Note: Previously called sfc32
```

Comment:
It might be nice to link to the source on Stack Overflow in this comment.

Reply:
Okay, I was thinking about that, but I wasn't sure how appropriate that would be. I'm aways thinking about things affecting cheating now 😅

Reply:
Now that you mention it... I suppose that a cheater could still get the correct answer from the shuffled options using something like the following procedure:

- List all permutations of the options listed in the `correct` field.
- For each permutation _P_, compute the hash that would result if that permutation were the original (correct) one. Then shuffle the options based on that hash.
- If the shuffled options match the `correct` value sent from the server, then _P_ is the correct answer!

To block this cheating vector, we would need the seed to be a random value that is stored in the database and _not_ revealed to the client. That would require a change to the Sorter editor to generate and store this seed. We would also have to backfill a seed for existing Sorter widgets.

Reply:
Add the resource to both functions that were borrowed

Reply:
Wow, that sounds like a lot, @benchristel. Is that something we need to do? That sounds like more work than just solving the problem XD

Reply:
It is more work, until someone makes a browser extension to automate it. That's how almost all cheating happens currently, AFAIK.

Maybe there's already some other piece of data which is stored in the database but not revealed to the client? If so, we could incorporate that into the seed to make it secure. I'm imagining something like the publish time of the content item. Though piping that info through to the options splitting code would be a pain.